src/vs/workbench/contrib/notebook/browser/contrib/notebookVariables/notebookInlineVariables.ts

```diff
@@ -0,0 +1,356 @@
+		const model = cell.textModel;
+		if (!model) {
+			return;
+		}
+
+		this.clearCellInlineDecorations(cell);
+		const inlineDecorations: IModelDeltaDecoration[] = [];
+
+
+		// might need to be a new provider registry, since we will have these that don't necessarily provide at the right timing
+		if (this.languageFeaturesService.inlineValuesProvider.has(model)) {
+			// use extension based provider, borrowed from https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugEditorContribution.ts#L679
+			const lastLine = model.getLineCount();
+			const lastColumn = model.getLineMaxColumn(lastLine);
+			const ctx: InlineValueContext = {
+				frameId: 0, // ignored, we won't have a stack from since not in a debug session
+				stoppedLocation: new Range(lastLine, lastColumn, lastLine, lastColumn) // executing cell by cell, so "stopped" location would just be the end of document
+			};
+
+			const providers = this.languageFeaturesService.inlineValuesProvider.ordered(model).reverse();
+			const lineDecorations = new Map<number, InlineSegment[]>();
+
+			const fullCellRange = new Range(1, 1, lastLine, lastColumn);
+
+			const promises = providers.flatMap(provider => Promise.resolve(provider.provideInlineValues(model, fullCellRange, ctx, CancellationToken.None)).then(async (result) => {
+				if (result) {
+
+					let kernel: INotebookKernelMatchResult;
+					const kernelVars: VariablesResult[] = [];
+					if (result.some(iv => iv.type === 'variable')) { // if anyone will need a lookup, get vars now to avoid needing to do it multiple times
+						if (!this.notebookEditor.textModel) {
+							return; // should not happen, a cell will be executed
+						}
+						kernel = this.notebookKernelService.getMatchingKernel(this.notebookEditor.textModel);
+						const variables = kernel?.selected?.provideVariables(event.notebook, undefined, 'named', 0, CancellationToken.None);
+						if (!variables) {
+							return;
+						}
+						for await (const v of variables) {
+							kernelVars.push(v);
+						}
+					}
+
+					for (const iv of result) {
+						let text: string | undefined = undefined;
+						switch (iv.type) {
+							case 'text':
+								text = (iv as InlineValueText).text;
+								break;
+							case 'variable': {
+								const name = (iv as InlineValueVariableLookup).variableName;
+								if (!name) {
+									continue; // skip to next var, no valid name to lookup with
+								}
+								const value = kernelVars.find(v => v.name === name)?.value;
+								if (!value) {
+									continue;
+								}
+								text = format('{0} = {1}', name, value);
+							}
+							case 'expression': {
+								let expr = (iv as InlineValueExpression).expression;
+								if (!expr) {
+									const lineContent = model.getLineContent(iv.range.startLineNumber);
+									expr = lineContent.substring(iv.range.startColumn - 1, iv.range.endColumn - 1);
+								}
+								if (expr) {
+									const expression = new Expression(expr);
```

Comment:
If I understand correctly, we don't have a way to evaluate expression with kernels.